### PR TITLE
Merge the latest changes to modules, examples & scripts for Lab Services libraries

### PR DESCRIPTION
**Pull request by petehauge**
**Added Scripts:**
- Assign-LabLiaisonRoleToLabAccounts.ps1
- Create-AadGroups.ps1
- Create-AzLabAccounts.ps1
- Create-AzLabs.ps1
- Get-LabMetrics.ps1
- Get-RegistrationLinks.ps1
- Import-LabLiaisonRole.ps1
- Publish-AzLabs.ps1
- Remove-AzLabAccounts.ps1
- Remove-AzLabs.ps1
- Reset-StudentAvailableHours.ps1
- Reset-StudentQuotasRunbook.ps1
- UploadImageToSharedImageGallery.ps1
- Validate-AzLabs.ps1

**Az.LabServices Module**
- Enable running in Azure Automation via grabbing token from globals
- Retries on a few conditions for InvokeRest method (always retry GET, retry on some return codes)
- Enable waiting on deletes (for removing labs)
- Enable connecting Shared Image Gallery by ID & enabling/disabling images for lab accounts
- Fix Set-AzLab to update changed properties (so we don’t accidently reset other properties)
- Fixes for New-AzLabSchedule to enable Stop-only schedules
- New Commands:
  - Get-AzLabAccountSharedGallery
  - Set-AzLabAccountSharedImage
  - Sync-AzLabADUsers
  - Start-AzLabStudentVm

**LabCreationLibrary Module**
- A ton of additional validation of the CSV file on import
- New Commands:
  - New-AzLabAccountsBulk
  - New-AzLabsBulk
  - Publish-AzLabsBulk
  - Set-AzRoleToLabAccountsBulk
  - Remove-AzLabsBulk
  - Sync-AzLabADUsersBulk
  - Get-AzLabsRegistrationLinkBulk
  - Reset-AzLabUserQuotaBulk
  - Convert-UsageQuotaToHours
  - Test-AzLabsBulk

### CODE REVIEW

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 [hunk 1]
```diff
@@ -251,14 +259,59 @@ function ConvertToUri($resource) {
...
+            $shouldRetry = $false
+            $result = Invoke-WebRequest -Uri $FullUri -Method $Method -Headers $authHeaders -Body $Body -UseBasicParsing 
+        }catch{
+            #if we have reach max number of calls, rethrow error no matter what it is
+            if($currentCallCount -eq $maxCallCount){
+                throw
+            }
+
+            #retry if Rest method is GET
+            if ($Method -eq 'Get'){
```
Comment (petehauge) [inline samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:288]: Consider not retrying for 401 (auth error)

Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:288]: Updated!

File: samples/ClassroomLabs/Modules/Library/Tools/Examples/Create-AzLabs.ps1 [hunk 2]
```diff
@@ -0,0 +1,38 @@
+[CmdletBinding()]
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/Examples/Create-AzLabs.ps1:1]: Done

File: samples/ClassroomLabs/Modules/Library/Tools/Examples/Reset-StudentQuotasRunbook.ps1 [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+}
+if (-not (Get-Command -Name "Reset-AzLabUserQuotaBulk" -ErrorAction SilentlyContinue)) {
+    Write-Error "Unable to find the LabCreationLibrary.psm1 module, please add to the Azure Automation account"
+}
+if (-not (Get-Command -Name "Start-ThreadJob" -ErrorAction SilentlyContinue)) {
+    Write-Error "Unable to find the ThreadJob Powershell module, please add to the Azure Automation account"
+}
+
+# Get the Access Token for this session
+$azureRmProfile = Invoke-Expression "[Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile"
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/Examples/Reset-StudentQuotasRunbook.ps1:51]: Found a way to do this all in one place in the library!

File: samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1 [hunk 4]
```diff
@@ -34,75 +34,1019 @@ function Import-LabsCsv {
...
+                        # For each completed job we write the result back to the appropriate Config object, using the "name" field to coorelate
+                        $jobName = $_.Name
+                        $jobState = $_.State
+                        $config = $ConfigObject | Where-Object {$_.ResourceGroupName -ieq $jobName.Split('+')[0] -and $_.LabAccountName -ieq $jobName.Split('+')[1] -and $_.LabName -ieq $jobName.Split('+')[2]}
+
+                        if (Get-Member -InputObject $config -Name LabResult) {
+                            $config.LabResult = $jobState
+                        }
+                        else {
+                            Add-Member -InputObject $config -MemberType NoteProperty -Name "LabResult" -Value $jobState
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:478]: Just tested and this works already!

[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:478]: The part that doesn't work is writing out to CSV and then back in...  looking into this now...

[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:478]: Fixed this by adding to the Import command (save the orig fields) and having a new export command "Export-LabsCsv"

File: samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1 [hunk 5]
```diff
@@ -34,75 +34,1019 @@ function Import-LabsCsv {
...
+            # Return the objects with an additional property on the result of the operation
+            return $ConfigObject
+        }
+
+        # New-AzLab-Jobs returns the config object with an additional column, we need to leave it on the pipeline
+        New-AzLab-Jobs   -ConfigObject $aggregateLabs 
+    }
+}
+
+function Set-AzRoleToLabAccountsBulk {
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:498]: Will add to our backlog

File: samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1 [hunk 6]
```diff
@@ -123,252 +1067,154 @@ function Publish-Labs {
...
 
-        function New-Accounts {
-            [CmdletBinding()]
-            param(
-                [parameter(Mandatory = $true, ValueFromPipeline = $true)]
-                [psobject[]]
-                $ConfigObject
-            )
+function Convert-UsageQuotaToHours {
+    [CmdletBinding()]
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:1172]: Done!

File: samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1 [hunk 7]
```diff
@@ -123,252 +1067,154 @@ function Publish-Labs {
...
+    $usage = [System.Xml.XmlConvert]::ToTimeSpan($RawTimeSpan)
+    return [math]::Ceiling($usage.TotalHours)
+}
 
-            $block = {
-                param($path, $ResourceGroupName, $LabAccountName)
 
-                Set-StrictMode -Version Latest
-                $ErrorActionPreference = 'Stop'
+function Test-AzLabsBulk {
```
[the remark this answers is not in the document]
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1:1184]: Done!

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 [hunk 8]
```diff
@@ -1576,7 +1822,10 @@ function New-AzLabSchedule {
         [Array] $WeekDays = @('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'),
 
         [parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true, HelpMessage = "Notes for the class meeting.")]
-        $Notes = ""
+        $Notes = "",
+
+        [parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true, HelpMessage = "Event type.")]
+        [System.String]$EventType = "Standard"
```
Comment (EMaher) [inline samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:1828]: Should we add a validate set attribute here?

**Review by EMaher — APPROVED** [review]

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 (hunk 8 above)
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:1828]: Yep!  Good idea, I'll add this

**Pull request merged**